### PR TITLE
fix: push_path and pull_path include empty directories

### LINKED ISSUE #968: pull_path doesn't retrieve empty directories
Per [this comment](https://github.com/canonical/operator/pull/960#issuecomment-1628001631), `Container.pull_path` doesn't retrieve empty directories. This seems like a bug, not a feature (Git's behaviour notwithstanding). From a quick look at the code, I think `push_path` will have a similar problem.

We probably need to change the `_list_recursive` helper to also yield directories (before recursing), and the callers test if it's a directory and create it, otherwise copy the file.

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
Instead of just recursively iterating through the remote/local files, also iterate through the directories as well, ensuring that they are created even if they don't contain any files.

Also create/update relevant test cases.

Fixes #968

### CODE REVIEW

File: ops/model.py [hunk 1]
```diff
@@ -2275,6 +2275,9 @@ def local_list(source_path: Path) -> List[pebble.FileInfo]:
             try:
                 for info in Container._list_recursive(local_list, source_path):
                     dstpath = self._build_destpath(info.path, source_path, dest_dir)
+                    if info.type is pebble.FileType.DIRECTORY:
```
Comment (benhoyt) [inline ops/model.py:2278]: It wasn't clear to me that enum values should be compared by identity (`is` vs `==`), but it looks like they should be -- TIL! For reference: https://docs.python.org/3/howto/enum.html#comparisons

File: test/test_model.py [hunk 2]
```diff
@@ -981,28 +981,32 @@ def __init__(self,
...
+        self.want_dirs = want_dirs or set()
 
 
 recursive_list_cases = [
     PushPullCase(
         name='basic recursive list',
         path='/',
         files=['/foo/bar.txt', '/baz.txt'],
-        want={'/foo/bar.txt', '/baz.txt'},
+        want={'/foo/bar.txt', '/baz.txt', '/foo'},
```
Comment (benhoyt) [inline test/test_model.py:1003]: Super nit, but I think it'd read more clearly for `/foo` to be listed first here. I realize it doesn't make a difference for set comparison, but it seems like it should come before `/foo/bar.txt`. Same below.

File: test/test_testing.py [hunk 3]
```diff
@@ -4595,16 +4597,14 @@ def test_pull(self):
     def test_pull_path(self):
         (self.root / "foo").mkdir()
         (self.root / "foo/bar").write_text("bar")
-        # TODO: pull_path doesn't pull empty directories
-        # https://github.com/canonical/operator/issues/968
-        # (self.root / "foobar").mkdir()
```
Comment (benhoyt) [inline test/test_testing.py:4600]: Nice catch!

**Review by benhoyt — APPROVED** [review]
Looks good to me, thanks! Just one super-nit comment.

**Review by weiiwang01 — APPROVED** [review]

**Pull request merged**